### ISSUE TITLE
WIP Proposal: Introduce pipelines / target groups

### ISSUE THREAD

**Issue by gkonterman** [fsharp/FAKE#1971]
I'm working with FAKE for a couple of weeks/months and the thing that 'bothers' me the most are the (hard) dependencies between targets. 

- First, it can be pretty difficult to read and manage these dependencies when you have a lot of targets (say 15+). 
- Second, it is hard to create optional dependencies (correct me if I am wrong) inside a dependency order. You usually end up with creating a lot of dummy or wrapper targets in order to achieve this, which causes a lot of boilerplate. I know FAKE has soft dependencies (https://fake.build/core-targets.html#Soft-dependencies) but this will not be always the solution and can be come very hairy
- And third, it can be hard to grasp the dependency order from the FAKE file alone (I know it is outputted but when you need to change it, it can be a puzzle).

I also would like to have the possibility to have no dependency at all on a target. A target itself has no dependencies so I can invoke it very easily as standalone / single target (without `-s`). This proposal with pipelines means that a target dependency is context sensitive: when it is part of a pipeline, only then the dependencies are set/active

**Wouldn't it be nice to introduce `pipelines` or `target groups`. This way you can be very explicit and it will be extremely readable.**

Example: I have three build configurations, a simple one (ie local) a parallel one and one for CI

Pseudo code 
```
My targets
- Bootstrap (optional)
- Clean (optional)
- Restore
- CompileLibs 
- CompilePlugins
- CompileApps
- Compile (dummy)
- Reset (optional)
- Pack
- Publish (optional)

Now I want to build 3 pipelines

Pipeline.create "Local-Build"
 ==> Restore
 ==> Compile
 ==> Reset
 ==> Pack

Pipeline.create "Parallel-Build"
 ==> Bootstrap
 ==> Clean
 ==> Restore
 ==> CompileLibs  (parallel)
 ==> CompilePlugins (parallel)
 ==> CompileApps (parallel)
 ==> Pack
 ==> Publish

Pipeline.create "CI-Build"
 ==> Bootstrap
 ==> Clean
 ==> Restore
 ==> Compile
 ==> Pack
 ==> Publish
```

As far as I know, it is pretty hard or cumbersome to create such build configuration with FAKE that still remains clean and readable, or am I wrong? Especially the optional targets like `Clean` and `Reset` are very hard to do.
Second, will this be feasible? According to me, only 1 pipeline has to be built, so for that matter it should be possible right? (IE no conflict dependencies between pipelines because only 1 is invoked)

So to summarise you will get two options:
- Target dependencies: always set, hard on a target (scope: global). 
- Pipeline dependencies: only set on a target when inside a pipeline (scope: local/pipeline). Targets itself are free of deps which make them very reusable

Alternative syntax:
```
Target.Group "LocalBuild" [ Restore, Compile, Pack ]
```

**Comment by matthid** [comment]
Yes I agree about the problem - this is indeed a frustrating point to deal with if you have lots of targets...

Regarding the solution:
Are you suggesting that "groups" are run in-order? The "advantage" of the current system is that we only specify dependencies and fake will figure out if we can run things in parallel. Would this still be possible with groups?

**Comment by matthid** [comment]
Sometimes it is a bit more difficult:

```fsharp
Target.Group "LocalBuild" [ 
  Restore, 
  [ Compile1, Compile2, CompileX ], <-- may run in parallel
  [ Pack1, Pack2, PackX ] 
]
```

You could run `Pack1` after `Compile1` is finished but fake still has to wait. This is the advantage of our "current" approach we can figure that out from the dependencies. Same goes for documentation generation which often can start alongside unit-testing or packaging

**Comment by matthid** [comment]
I think the problem is finding a syntax where you can "keep" the overall view (like you propose) but still allow to opt into the advantages from a more declarative approach.

(Also Brainstorming:)

```fsharp
Target.group "G1" { restore ==> compile1 ==> pack1 }

Target.group "LocalBuild" { "G1" <+> "G2" <+> ... } // or some kind of list?

```

**Comment by jindraivanek** [comment]
Idea: group pickup only dependencies that is between targets in that group. So

```
Target.Group "LocalBuild" [ Restore, Compile, Reset, Pack ]
```

would not run `Bootstrap` and `Clean` but still pickup order (and parallelization) of its targets from dependencies.

In other words, group would say "ignore existence of other targets (not in this group)".

**Comment by agross** [comment]
My perspective as a Ruby/rake user: a target or task does not necessarily need to perform actions. In Ruby you would define such groups just by adding tasks (roughly equivalent to targets) missing the actual action. This makes it easy to define such meta-tasks that you describe.

**Comment by gkonterman** [comment]
> (Also Brainstorming:)
> 
> Target.group "G1" { restore ==> compile1 ==> pack1 }
> 
> Target.group "LocalBuild" { "G1" <+> "G2" <+> ... } // or some kind of list?

Yes I was thinking about the same. But I remember that FAKE does not want any more special 'cryptic' operators (and I agree)

**Comment by BlythMeister** [comment]
In my company build tools which we are re-working for fake we have this issue.
We do this:
https://github.com/15below/Build.Tools/blob/feature/FAKE5/BuildTypes.fsx
Defines the pipe lines
https://github.com/15below/Build.Tools/blob/feature/FAKE5/Targets.fsx
Defines the targets and sub trees for the pipeline
It's nasty to work with. But throwing a use case into the mix 🙂

**Comment by gkonterman** [comment]
@BlythMeister Wow, that's actually very nice! I can get much inspiration from this 👍 

Remains the question: should FAKE provide first class support for this?

**Comment by BlythMeister** [comment]
@gkonterman ahh thanks.  The fake5 feature is a WIP as I upgrade our tooling.

Personally, I think that it's not required in FAKE.
one of the things I like about fake is the ability to do whatever you like in the script.

**Comment by haf** [comment]
One of the reasons that Vesa wrote Hopac was to build a parallel build tool. https://www.amazon.com/Concurrent-Programming-ML-John-Reppy/dp/0521480892/ref=sr_1_1?ie=UTF8&qid=1527882421&sr=8-1&keywords=concurrent+programming+ml contains a sample implementation of such a system. Seems like FAKE could be using it?

Furthermore, if FAKE ever would aspire to say, join where the rest of the world is going; towards https://bazel.build/ and the like, and help the .net devs build really quick builds with distributed caching and such, it would be great if it could wrap the F# compiler rather than relying on msbuild for happy-path projects (i.e. netcoreapp2.0/netstandard2.0-only with all-net-core dependencies).

**Comment by matthid** [comment]
> Furthermore, if FAKE ever would aspire to say, join where the rest of the world is going

@haf That might be a bit exaggerated but yes I'm not against working in a direction like this. However we would need someone to do the work and I'd like to stabilize and improve the runtime first. I guess we would realize this in a new module rather than trying to change the current `fake.core.target` module?

So this issue in particular is about the problem of "managing" lots of targets and in particular running targets in different scenarios. 

@agross Yes, in fact we already do that by using "empty" targets (`ignore` function) but it is a bit unsatisfying at the moment.

I kind of like the suggestion of @jindraivanek . You still need to define the dependencies but you might need less "meta" and "empty" targets to have the behavior you want.

**Comment by gkonterman** [comment]
> I kind of like the suggestion of @jindraivanek . You still need to define the dependencies but you might need less "meta" and "empty" targets to have the behavior you want.

> In other words, group would say "ignore existence of other targets (not in this group)".

I think we can agree that this is the basic feature we want. A way to create a group of targets, which trims all the existing (hard) dependencies.

So now we need to agree upon the names and syntax i guess? Personally I think `Target.group` fits the best in the existing API (also Paket uses groups, so its a 'known' keyword).

**Comment by gkonterman** [comment]
# Naming

## `Target.group "LocalBuild"`

Please vote 👍 or 👎

**Comment by gkonterman** [comment]
# Naming

## `Pipeline.create "LocalBuild"`

Please vote 👍 or 👎

**Comment by gkonterman** [comment]
# Naming

## `BuildType.create "LocalBuild"`

Please vote 👍 or 👎

**Comment by gkonterman** [comment]
# Syntax

## `<CreateFunction> [ Restore, Compile, Test, Pack ]`

Aka list syntax

Please vote 👍 or 👎

**Comment by gkonterman** [comment]
# Syntax

## `<CreateFunction> Restore =|> Compile =|> Test =|> Pack`

Aka custom operator (arrow) syntax
PS: type of arrow can be changed, Im not sure we can use `==>` or `=|>` (because one is already used for hard deps, and the other is not possible?)

Please vote 👍 or 👎

**Comment by jackfoxy** [comment]
FAKE is the first and only build DSL I have used, so from my perspective making it a DSL using the conventions of previous build DSLs was not useful for me. I don't find the convention of targets useful, and coming from thinking in F#, confusing. Most targets seem to be () -> () signature functions with side effects, and others are just end of pipeline statements.

I would like to see FAKE have well-documented alternatives to all the custom operators, maybe even including the globbing operators. (This is probably already mostly possible, just not emphasized in the documentation or existing implementations.) I've gotten used to the DSL build convention over time, but wouldn't it make more sense to make build scripts like we make every other F# program? Targets defined as a discriminated union, the bottom of the build file is a match statement, and each target is preceded by a series of |>, making the dependencies transparent.

CI services like Appveyor and Travis exist in large part to tame the wilderness of side effects. Maybe there is a way forward with Paket, FAKE, dotnet CLI, and F#.Formatting. Pure functions are to be favored, side effects isolated and transparent, custom operators avoided. For backwards compatibility the DSL mimicing other make systems will have to remain, but it should not be the canonical way.

**Comment by gkonterman** [comment]
> Targets defined as a discriminated union, the bottom of the build file is a match statement, and each target is preceded by a series of |>, making the dependencies transparent.

I like that idea, thats also why I use `=|>` because I like the pipe idea (especially in this context). Being able to use just 'TargetA |> TargetB' would be nice imo. Not sure what this means for FAKE in terms of rewriting a lot of the core

**Comment by matthid** [comment]
Would we run pipelines via CLI the same way we run targets?

`fake run build.fsx target mygroup`

Or would we extend the CLI like 

`fake run build.fsx group mygroup`

?

**Comment by jackfoxy** [comment]
What does =|> do? Part of the problem I have with operators is looking up their documentation.

**Comment by gkonterman** [comment]
> What does =|> do? 

Creating a dependency between targets in a group context, but maybe it is possible to use the default `==>`

> Part of the problem I have with operators is looking up their documentation.

Agree, I think that every operator should always have a "textual" counterpart, ie you can use `==>` or the function `Target.HasDependency`. So a short and a long (readable) notation. 
In fact, I think you should always start with the named function and add a custom operator on top if it is very clear and obvious/common (like `|>`). But I have some background in Scala and I never liked custom operators. Although in some situations it can be quite nice, it remains a "dangerous weapon".

> Or would we extend the CLI like
> fake run build.fsx group mygroup ?

I like `fake run build.fsx group mygroup`

**Comment by vbfox** [comment]
I like the idea of having some ordered groups that could replace optional dependencies for ordering purposes.

The syntax I have now (Uses my [BuildTask](https://www.nuget.org/packages/BlackFox.Fake.BuildTask/) lib ) look like that:

```fsharp
let bootstrap = BuildTask.create "Bootstrap" [] { (* ... *) }
let clean = BuildTask.create "Clean" [bootstrap.IfNeeded] { (* ... *) }
let restore = BuildTask.create "Restore" [clean.IfNeeded] { (* ... *) }
let compileLibs = BuildTask.create "CompileLibs" [restore] { (* ... *) }
let compilePlugins = BuildTask.create "CompilePlugins" [restore] { (* ... *) }
let compileApp = BuildTask.create "CompileApp" [restore] { (* ... *) }
let compile = BuildTask.createEmpty "Compile" [compileLibs; compilePlugins; compileApp]
let reset = BuildTask.create "Reset" [compile.IfNeeded] { (* ... *) }
let pack = BuildTask.create "Pack" [reset.IfNeeded; compile] { (* ... *) }
let publish = BuildTask.create "Publish" [pack] { (* ... *) }

let localBuild = BuildTask.createEmpty "LocalBuild" [reset; pack]
let parallelBuild = BuildTask.createEmpty "ParallelBuild" [bootstrap; clean; publish]
let ciBuild = BuildTask.createEmpty "CIBuild" [bootstrap; clean; publish]
```

**Comment by matthid** [comment]
I have added 
![image](https://user-images.githubusercontent.com/1866463/46903381-42913000-ced4-11e8-9f34-d8cd280cf42a.png)

I think it is a good and much more lightweight way to do this. However one limitation of this is if you dynamically create targets this doesn't work easily. Then you probably need to understand the underlying infrastructure...

I'm closing this for now, but new comments/PRs are welcome. We can reopen this if people get interested into this again. (or start a new discussion)

**Comment by inosik** [comment]
I'd be interested in something like this as well. I often do something like this (using the task library by @vbfox):

``` fsharp
let buildTests = BuildTask.createFn "Build tests" [] ...
let runTests = BuildTask.createFn "Run Tests" [] ...
let test = BuildTask.createEmpty "Test" [buildTests; runTests]

let buildApp = BuildTask.createFn "Build app" [] ...
let generateResources = BuildTask.createFn "Generate resources" [] ...
let build = BuildTask.createEmpty "Build" [buildApp; generateResources]

let defaultT = BuildTask.createEmpty "Default" [test; build]
```

I'd like to be able to "group" the dependencies of "Test" and "Build" in a way so that none of the tasks in the "Build" group are run until all the tasks in the "Test" group are executed. Right now, "Build app" might be executed before "Run tests", which doesn't make sense. This is hard to do at the moment without doing something like @vbfox has shown in [his comment](https://github.com/fsharp/FAKE/issues/1971#issuecomment-415051570), which makes the script really hard to read.
